feat: Document dynamic sampling (#2954)
getsentry/sentry-docs: src/docs/product/data-management-settings/dynamic-sampling/configuration.mdx

---
title: Dynamic Sampling Configuration
sidebar_order: 40
description: "Learn more about how dynamic sampling rules are applied."
---

> As described in [sampling types](/product/data-management-settings/dynamic-sampling/sampling-types/) dynamic sampling
> operates on both error events and transaction events, with transaction events having the ability to be sampled
> independently or as part of a complete trace.
> 

Reviewer: ```suggestion
Dynamic sampling operates on both error events and transaction events, with transaction events having the ability to be sampled independently or as part of a complete trace.

## Individual Errors and Transactions

Error and individual transaction sampling decisions are made based on the event data.

A sampling rule is selected from the available sampling rules by going through the list of dynamic sampling rules for
errors or transactions and picking up [the first rule that matches the event data.](#rule-selection) 

Individual transaction rules are based on the transaction attributes and not on the trace context attributes and
therefore can target individual services that are not initiating traces.

While sampling transaction traces gives a very good view of how transactions flow through the system, sometimes we are
interested in analysing transactions from a particular service that happens not to be the initiator of transaction
(e.g. a backend server). If we are interested in creating rules based on attributes specific to our service, for
example we have a suspicion that a particular release might have some performance problems, we can sample based on
transaction event.

```

> To configure dynamic sampling you need to have organization owner or manager role.
> 
> The dynamic sampling configuration can be found in the project settings under "Filters & Sampling".
> You can find two lists one for error rules and one for transaction rules. The transaction rules list contains
> both rules for independent transactions and rules for transaction traces.
> 

Reviewer: ```suggestion
```

> Regardless of the type of rule the basic structure of the rule is the same:
> 
> * has a type that determines the event type to which it applies (Error or Transaction).
> * has 0 or more conditions
> * has a sampling rate (in percentages)
> 
> Transaction rules are further divided in, independent transaction rules and trace transaction rules.
> 
> If a rule contains multiple conditions the conditions will have to be met simultaneously (`and` logical operation) in
> order to have the rule applied.
> 
> A condition may contain multiple values to be matched, for the condition to be matched any of values must match
> (`or` logical operation).

Reviewer: ```suggestion
```

## Transaction trace rules

> While dynamic sampling rules for errors or independent transactions have a variety of conditions that can be selected
> transaction traces are a lot more limited in the information that can be used for selection.
> 
> When sampling transaction traces, dynamic sampling rules are based on the transaction context.
> This means that dynamic sampling decisions are based on information extracted from the system that started the transaction.
> Transaction trace dynamic sampling rules are based on the following attributes,
> all of which belong to the transaction that started the trace:
> 
> - Project ID
> - Release
> - Environment
> - User segment
> - Transaction name

Reviewer: ```suggestion

Sentry offers the possibility to sample transactions either independently or as part of a trace. Sampling transactions using transaction traces is advantageous when trying to understand transactions in the context of
the global system. In this case we are not interested in seeing a particular transaction, but the whole group of
transactions in a trace so sampling should occur at the trace level.

Learn more about tracing in our [full Tracing documentation](/product/sentry-basics/tracing/). 

While dynamic sampling rules for errors or independent transactions have a variety of conditions that can be selected
transaction traces are a lot more limited in the information that can be used for selection.

When sampling transaction traces, dynamic sampling rules are based on the transaction context.
This means that dynamic sampling decisions are based on information extracted from the system that started the transaction.
Transaction trace dynamic sampling rules are based on the following attributes,
all of which belong to the transaction that started the trace:

- Project ID
- Release
- Environment
- User segment
- Transaction name
```

## Dynamic Sampling and Inbound Filters

Inbound filters are a related concept, with key differences from Dynamic Sampling.

Inbound filters may be viewed as a coarse sampling technique, where only 0% and 100% sampling rates are available.
The key difference between filters and sampling is that filters compose while sampling rules do not, the example
provides an illustration:

With two filters defined - one that removes all requests coming from Internet Explorer 9 or lower, and one that filters
requests coming from an application with release 1.* - all  events coming either from Explorer 9 or from a client
with release 1.* will be filtered out, effectively the two filters are ORed together.

With two sampling rules defined - first on release 1.* at 10%, and the second sampling on "prod1" environment
at 20% - the combined effect will not be  similar to the inbound filters case. A request coming for release 1.1 and
prod1 will match the first rule and will be sampled at 10% , a request for release 1.1 & prod2 will still match the
first rule and will be sampled at 10% while a request for release 2.1 and prod1 will match the second rule and will be
sampled at 20%.

> ## Rule selection

Reviewer: ```suggestion
## Rule Selection
```
